daemon/daemon.go

```diff
@@ -129,85 +116,50 @@ var polkitCheckAuthorization = polkit.CheckAuthorization
-		logger.Panicf("Command can't have RootOnly together with any *OK flag")
-	}
-
-	if user != nil && !c.RootOnly {
-		// Authenticated users do anything not requiring explicit root.
-		return accessOK
-	}
-
-	// isUser means we have a UID for the request
-	isUser := false
-	pid, uid, socket, err := ucrednetGet(r.RemoteAddr)
-	if err == nil {
-		isUser = true
-	} else if err != errNoID {
-		logger.Noticef("unexpected error when attempting to get UID: %s", err)
-		return accessForbidden
-	}
-	isSnap := (socket == dirs.SnapSocket)
-
-	// ensure that snaps can only access SnapOK things
-	if isSnap {
-		if c.SnapOK {
-			return accessOK
-		}
-		return accessUnauthorized
-	}
-
-	// the !RootOnly check is redundant, but belt-and-suspenders
-	if r.Method == "GET" && !c.RootOnly {
-		// Guest and user access restricted to GET requests
-		if c.GuestOK {
-			return accessOK
+func (c *Command) accessRules() []accessChecker {
+	if c.RootOnly {
+		if c.UserOK || c.GuestOK || c.SnapOK {
+			// programming error
+			logger.Panicf("Command can't have RootOnly together with any *OK flag")
 		}
-
-		if isUser && c.UserOK {
-			return accessOK
+		return []accessChecker{
+			denySnapSocket{},
+			allowRoot{},
 		}
 	}
-
-	// Remaining admin checks rely on identifying peer uid
-	if !isUser {
-		return accessUnauthorized
+	rules := []accessChecker{
+		allowSnapUser{},
 	}
-
-	if uid == 0 {
-		// Superuser does anything.
-		return accessOK
+	if c.SnapOK {
+		rules = append(rules, allowSnapSocket{})
+	} else {
+		rules = append(rules, denySnapSocket{})
+	}
+	if c.GuestOK {
+		rules = append(rules, allowGetByGuest{})
+	} else if c.UserOK {
+		rules = append(rules, allowGetByUser{})
 	}
+	rules = append(rules, allowRoot{})
+	if c.PolkitOK != "" {
+		rules = append(rules, polkitCheck{c.PolkitOK})
+	}
+	return rules
+}
 
-	if c.RootOnly {
-		return accessUnauthorized
+// canAccess checks if the command allows the given request.
+func (c *Command) canAccess(r *http.Request, user *auth.UserState) accessResult {
+	ucred, err := ucrednetGet(r.RemoteAddr)
+	if err != nil && err != errNoID {
+		logger.Noticef("unexpected error when attempting to get UID: %s", err)
+		return accessForbidden
 	}
 
-	if c.PolkitOK != "" {
-		var flags polkit.CheckFlags
-		allowHeader := r.Header.Get(client.AllowInteractionHeader)
-		if allowHeader != "" {
-			if allow, err := strconv.ParseBool(allowHeader); err != nil {
-				logger.Noticef("error parsing %s header: %s", client.AllowInteractionHeader, err)
-			} else if allow {
-				flags |= polkit.CheckAllowInteraction
-			}
-		}
-		// Pass both pid and uid from the peer ucred to avoid pid race
-		if authorized, err := polkitCheckAuthorization(pid, uid, c.PolkitOK, nil, flags); err == nil {
-			if authorized {
-				// polkit says user is authorised
-				return accessOK
-			}
-		} else if err == polkit.ErrDismissed {
-			return accessCancelled
-		} else {
-			logger.Noticef("polkit error: %s", err)
+	for _, rule := range c.accessRules() {
```

Comment:
because of the case where we get an  accessUnauthorized  this is still order dependent, it makes it more fragile than the current code. If we want to go in this direction I think we need to make the order of the rules irrelevant for that this should work more like:

* if some rules return OK and all the other returns unknown return OK
* if a rule returns Unauhorized return immediately with that
* if all rules returned Unknown return Unauhorized as well

Reply:
Ordering is important for the polkit rule, as it potentially has the side effect of prompting the user for their password.  We only want to do that as a last resort if no other checkers grant access.

Based on last week's conversations where you said it would be better to make the checks explicitly procedural if ordering is important.  Looking at the existing REST endpoints, I think these could be split into three separate checkers:

1. Optionally allow GET requests from everyone, and require one of macaroon, root, or polkit for other methods.  Requests from `snapd-snap.socket` are denied.
2. Something equivalent to the current `RootOnly` behaviour.
3. Something equivalent to the current `SnapOK` behaviour.

The checker for the theme API could slot in next to these three.

Reply:
That proposal is not too dissimilar to how we describe/classify Access here:

https://forum.snapcraft.io/t/snapd-rest-api/17954

as either  open, root, authenticated.   SnapOK is a bit orthogonal to the rest so we need to be careful.

In case we are looking for defaults (not sure):

* GET should default to authenticated
* everything else if we go with a default should be root-only

Reply:
I opted not to have a default access policy, and instead added a test to make sure every command has a non-nil access checker corresponding to the HTTP methods it implements.